desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,818 @@
+    return T, B
+
+
+def _swap_pl(f):
+    # Given shape (L, num_pitch, -1) or (num_pitch, L, -1) or (num_pitch, -1)
+    # swap L and num_pitch axes.
+    assert f.ndim <= 3
+    return jnp.swapaxes(f, 0, -2)
+
+
+# TODO: After GitHub issue #1034 is resolved, we should pass in the previous
+#  θ(α, ζ) coordinates as an initial guess for the next coordinate mapping.
+#  Perhaps tell the optimizer to perturb the coefficients of the
+#  θ(α, ζ) directly? think this is equivalent to perturbing lambda.
+
+
+class Bounce2D(IOAble):
+    """Computes bounce integrals using two-dimensional pseudo-spectral methods.
+
+    The bounce integral is defined as ∫ f(λ, ℓ) dℓ, where
+        dℓ parameterizes the distance along the field line in meters,
+        f(λ, ℓ) is the quantity to integrate along the field line,
+        and the boundaries of the integral are bounce points ℓ₁, ℓ₂ s.t. λ|B|(ℓᵢ) = 1,
+        where λ is a constant defining the integral proportional to the magnetic moment
+        over energy and |B| is the norm of the magnetic field.
+
+    For a particle with fixed λ, bounce points are defined to be the location on the
+    field line such that the particle's velocity parallel to the magnetic field is zero.
+    The bounce integral is defined up to a sign. We choose the sign that corresponds to
+    the particle's guiding center trajectory traveling in the direction of increasing
+    field-line-following coordinate ζ.
+
+    Notes
+    -----
+    Brief description of algorithm.
+
+    Magnetic field line with label α, defined by B = ∇ρ × ∇α, is determined from
+        α : ρ, θ, ζ ↦ θ + λ(ρ,θ,ζ) − ι(ρ) [ζ + ω(ρ,θ,ζ)]
+    Interpolate Fourier-Chebyshev series to DESC poloidal coordinate.
+      θ : α, ζ ↦ tₘₙ exp(jmα) Tₙ(ζ)
+    Compute |B| along field lines.
+      |B| : α, ζ ↦  bₙ(θ(α, ζ)) Tₙ(ζ)
+    Compute bounce points.
+      r(ζₖ) = |B|(ζₖ) − 1/λ = 0
+    Interpolate smooth components of integrand with FFTs.
+      G : α, ζ ↦ gₘₙ exp(j [m θ(α,ζ) + n ζ] )
+    Perform Gaussian quadrature after removing singularities.
+      Fᵢ : λ, ζ₁, ζ₂ ↦  ∫ᵢ f(λ, ζ, {Gⱼ}) dζ
+
+    Longer description for developers.
+
+    For applications which reduce to computing a nonlinear function of distance
+    along field lines between bounce points, it is required to identify these
+    points with field-line-following coordinates. (In the special case of a linear
+    function summing integrals between bounce points over a flux surface, arbitrary
+    coordinate systems may be used as that task reduces to a surface integral,
+    which is invariant to the order of summation).
+
+    The DESC coordinate system is related to field-line-following coordinate
+    systems by a relation whose solution is best found with Newton iteration.
+    There is a unique real solution to that relation, so Newton iteration is a
+    globally convergent root-finding algorithm here. For the task of finding
```

Comment:
This is the map from (theta, zeta) to (alpha, zeta) right?

not sure its formally globally convergent, just because it has a unique root?

Reply:
Hmm, it's a unique root and the derivative is never flat; I suppose I didn't consider whether it can get stuck oscillating

Reply:
I'll just say unique root, because I shouldn't claim it converges